src/workerd/server/server.c++

```diff
@@ -1373,23 +1379,235 @@ public:
+
+      bool hasClients() { return containerRef != kj::none; }
+      kj::Maybe<ActorContainerRef&> getContainerRef() { return containerRef; }
+
+      // The actor is constructed after the ActorContainer so it starts off empty.
+      kj::Maybe<kj::Own<Worker::Actor>> actor;
+    private:
+      kj::StringPtr key;
+      kj::Own<RequestTracker> tracker;
+      ActorNamespace& parent;
+      kj::Timer& timer;
+      kj::TimePoint lastAccess;
+      kj::Maybe<kj::Own<Worker::Actor::HibernationManager>> manager;
+      kj::Maybe<kj::Promise<void>> shutdownTask;
+
+      // Non-empty if at least one client has a reference to this actor.
+      // If no clients are connected, we may be evicted by `cleanupLoop`.
+      kj::Maybe<ActorContainerRef&> containerRef;
+      friend class ActorContainerRef;
+    };
+
+    // This class tracks clients that a have reference to the given actor.
+    // Upon destruction, we update the lastAccess time for the actor and
+    // `ActorContainer::hasClients()` starts returning false. After 70 seconds, the cleanupLoop
+    // will remove the `ActorContainer` from `actors`.
+    class ActorContainerRef: public kj::Refcounted {
+    public:
+      ActorContainerRef(ActorContainer& container): container(container) {
+        // Link this ref to the actual ActorContainer.
+        container.containerRef = *this;
+      }
+      ~ActorContainerRef() {
+        KJ_IF_SOME(ref, container) {
+          ref.updateAccessTime();
+          ref.containerRef = kj::none;
+        }
+      }
+
+      kj::Own<ActorContainerRef> addRef() {
+        return kj::addRef(*this);
+      }
+
+    private:
+      // This is a maybe because the ActorContainer could be destroyed before ActorContainerRef
+      // if the actor is broken.
+      kj::Maybe<ActorContainer&> container;
+      friend class ActorContainer;
+    };
+
   private:
     WorkerService& service;
     kj::StringPtr className;
     const ActorConfig& config;
-    kj::HashMap<kj::String, kj::Own<Worker::Actor>> actors;
-    kj::TaskSet onBrokenTasks;
+    // If the actor is broken, we remove it from the map. However, if it's just evicted due to
+    // inactivity, we keep the ActorContainer in the map but drop the Own<Worker::Actor>. When a new
+    // request comes in, we recreate the Own<Worker::Actor>.
+    kj::HashMap<kj::String, kj::Own<ActorContainer>> actors;
+    kj::HashMap<kj::String, kj::Maybe<kj::Promise<void>>> onBrokenTasks;
+    kj::Maybe<kj::Promise<void>> cleanupTask;
+    kj::Timer& timer;
 
     kj::Promise<kj::Own<WorkerInterface>> getActorThenStartRequest(
         kj::String id,
         IoChannelFactory::SubrequestMetadata metadata) {
-      kj::Own<Worker::Actor> actor = co_await getActorImpl(kj::mv(id));
-      co_return service.startRequest(kj::mv(metadata), className, kj::mv(actor));
+      auto pair = co_await getActorImpl(kj::mv(id));
+      auto actor = kj::mv(pair.first);
+      auto refTracker = kj::mv(pair.second);
```

Comment:
```suggestion
      auto [actor, refTracker] = co_await getActorImpl(kj::mv(id));
```

Note that this works with either `std::pair` or a struct (as long as the struct's fields are defined in that order).

Reply:
Thanks for the tip!